hal-x86_64/src/mm/mod.rs

```diff
@@ -0,0 +1,831 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        struct FmtFlags<'a, L>(&'a Entry<L>);
+        impl<'a, L> fmt::Debug for FmtFlags<'a, L> {
+            #[inline(always)]
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                macro_rules! write_flags {
+                    ($f:expr, $e:expr, $($bit:ident),+) => {
+                        let mut wrote_any = false;
+                        $(
+                            if $e & Entry::<L>::$bit != 0 {
+                                write!($f, "{}{}", if wrote_any { " | " } else { "" }, stringify!($bit))?;
+                                // the last one is not used but we can't tell the macro that easily.
+                                #[allow(unused_assignments)]
+                                {
+                                    wrote_any = true;
+                                }
+                            }
+                        )+
+                    }
+                }
+
+                write_flags! {
+                    f, self.0.entry,
+                    PRESENT,
+                    WRITABLE,
+                    WRITE_THROUGH,
+                    CACHE_DISABLE,
+                    ACCESSED,
+                    DIRTY,
+                    HUGE,
+                    GLOBAL
+                }
+                Ok(())
+            }
+        }
+        f.debug_struct("Entry")
+            .field("level", &format_args!("{}", L::NAME))
+            .field("addr", &self.phys_addr())
+            .field("flags", &FmtFlags(&self))
+            .finish()
+    }
+}
+
+pub mod size {
+    use super::Size;
+
+    #[derive(Copy, Clone, Eq, PartialEq)]
+    pub enum Size4Kb {}
+
+    impl Size for Size4Kb {
+        const SIZE: usize = 4 * 1024;
+        const PRETTY_NAME: &'static str = "4KB";
+    }
+
+    #[derive(Copy, Clone, Eq, PartialEq)]
+    pub enum Size2Mb {}
+
+    impl Size for Size2Mb {
+        const SIZE: usize = Size4Kb::SIZE * 512;
+        const PRETTY_NAME: &'static str = "2MB";
+    }
+
+    #[derive(Copy, Clone, Eq, PartialEq)]
+    pub enum Size1Gb {}
+
+    impl Size for Size1Gb {
+        const SIZE: usize = Size2Mb::SIZE * 512;
+
+        const PRETTY_NAME: &'static str = "1GB";
+    }
+}
+
+const RECURSIVE_INDEX: usize = 0o777;
+const SIGN: usize = 0o177777 << 48;
+pub trait Level {
+    const NAME: &'static str;
+    const SUBLEVELS: usize;
+
+    const INDEX_SHIFT: usize = 12 + (9 * Self::SUBLEVELS);
+
+    fn index_of(v: VAddr) -> usize {
+        (v.as_usize() >> Self::INDEX_SHIFT) & RECURSIVE_INDEX
+    }
+
+    fn table_addr(v: VAddr) -> VAddr;
+}
+
+impl<L: Level> fmt::Debug for PageTable<L> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("PageTable")
+            .field("level", &format_args!("{}", L::NAME))
+            .field("addr", &format_args!("{:p}", self))
+            // .field("entries", &self.entries)
```

Comment:
Can we drop this comment?

It might also be cool to have this eventually print all non-empty entries, but we don't need to do that yet.